File: algorithms/sort/bubble_sort.js
const bubbleSort = (nums) => {
  let isSorted = false;

  while (!isSorted) {
    isSorted = true;
    for (let i = 0; i < nums.length - 1; i++) {
      if (nums[i] > nums[i+1]) {
        swap(nums, i, i+1);
        isSorted = false;
      }
    }
  }
};

const swap = (nums, i, j) => {
  const temp = nums[i];
  nums[i] = nums[j];
  nums[j] = temp;
};

/* Test */
const numbers = [9, 3, 5, 2, 6, 24, 52, 90, 43, 5];
bubbleSort(numbers);
console.log(numbers);

File: Python/algorithms/sort_and_search/bubble_sort.py
#! /usr/bin/python
"""
Time: O(N^2) time
Space: O(1) space
In-place
Stable
"""
def bubbleSort(nums):
    isSorted = False
    lastUnsortedIndex = len(nums)-1
    while not isSorted:
        isSorted = True
        for i in range(lastUnsortedIndex):
            if nums[i] > nums[i+1]:
                nums[i], nums[i+1] = nums[i+1], nums[i]
                isSorted = False
        lastUnsortedIndex -= 1


if __name__ == "__main__":
    sample = [3, 1, 4, 5, 10, -4, 2, -5]
    print("Before sorting", sample)
    bubbleSort(sample)
    print("After sorting", sample)
